net/basic_socket.h

```diff
@@ -101,36 +105,48 @@ __FORCE_INLINE__ int doio(IOCB iocb, WAIT waitcb) {
     }
 }
 
-template <typename IOCB>
-__FORCE_INLINE__ ssize_t doio_n(void *&buf, size_t &count, IOCB iocb) {
-    ssize_t n = 0;
-    while (count > 0) {
-        ssize_t ret = iocb();
-        if (ret < 0) return ret; // error
-        if (ret == 0) break; // EOF
-        (char *&)buf += ret;
-        count -= ret;
+#define DOIO_ONCE(iocb, waitcb) doio_once(LAMBDA(iocb), LAMBDA(waitcb))
+
+template <typename IOCB, typename ADV> __FORCE_INLINE__
+ssize_t doio_loop(IOCB iocb, ADV advance) {
+    ssize_t ret, n = 0;
+    do {
+        ret = iocb();
+        if (ret <  0) return ret; // error
+        if (ret == 0) break;  // EOF
         n += ret;
-    }
+    } while (advance(ret, n));
     return n;
 }
 
-template <typename IOCB>
-__FORCE_INLINE__ ssize_t doiov_n(iovector_view &v, IOCB iocb) {
-    ssize_t count = 0;
-    while (v.iovcnt > 0) {
-        ssize_t ret = iocb();
-        if (ret < 0) return ret;
-        if (ret == 0) break;
-        count += ret;
-
-        uint64_t bytes = ret;
-        auto extracted = v.extract_front(bytes);
-        assert(extracted == bytes);
+struct BufAdv {
```

Comment:
Adv looks like `add v` but the actual meaning we are going to express is `advance` ?

Reply:
Yes, and better naming is welcome.

Reply:
BufAdvance

Reply:
how about step?